src/main/webapp/cluster-view/cluster-view.controller.js

```diff
@@ -1,26 +1,150 @@
+      };
+
+      $scope.currentClusterHasCaches = function () {
+        return !ispn.util.isEmptyObject($scope.currentCluster.getCaches());
+      };
+
+      $scope.isNodesView = function () {
+        return $scope.shared.currentCollection === 'nodes';
+      };
+
+      $scope.isCachesView = function () {
+        return $scope.shared.currentCollection === 'caches';
+      };
+
+      $scope.zoomValue = 0;
+
+      $scope.isZoomedIn = function () {
+        return $scope.zoomValue == 0 ? false : true;
+      };
+
+      $scope.initialMin = 0;
+      $scope.initialMax = 1000;
+      $scope.sliders = [
+        {
+          name: "Average read time",
+          id: 'average-read-time',
+          from: $scope.initialMin,
+          to: $scope.initialMax
+        },
+        {
+          name: "Average write time",
+          id: 'average-write-time',
+          from: $scope.initialMin,
+          to: $scope.initialMax
+        },
+        {
+          name: "Average remove time",
+          id: 'average-remove-time',
+          from: $scope.initialMin,
+          to: $scope.initialMax
+        },
+        {
+          name: "Average replication time",
+          id: 'average-replication-time',
+          from: $scope.initialMin,
+          to: $scope.initialMax
+        }
+      ];
+
+      angular.forEach($scope.sliders, function (slider) {
+        $scope.$watch(function () {
+          return slider;
+        }, function (newValue, oldValue) {
+          //even though newValue is the right slider that has changed
+          //we still have to check all sliders
+          if ($scope.isCachesView()) {
+            filterOutCaches($scope.sliders);
+          } else if ($scope.isNodesView()) {
+            filterOutNodes($scope.sliders);
+          }
+        }, true);
+      });
+
+      $scope.clearFilters = function () {
+        angular.forEach($scope.sliders, function (slider) {
+          slider.from = $scope.initialMin;
+          slider.to = $scope.initialMax;
+        });
+      };
+
+      var filterOutCaches = function (sliderArray) {
+        var cachesMap = $scope.currentCluster.getCaches();
+        angular.forEach(cachesMap, function (cache) {
+          var p = modelController.getServer().fetchCacheStats($scope.currentCluster, cache);
+          p.then(function (response) {
+            return responseWithinAllSliderRanges(sliderArray, response[0]);
+          }).then(function (result) {
+            cache.show = result;
+          });
+        });
+      };
+
+      var filterOutNodes = function (sliderArray) {
+        var server = modelController.getServer();
+        var nodes = server.getNodes();
+        var promises = [];
+        angular.forEach(nodes, function (node) {
+          promises.push(server.fetchAggregateNodeStats($scope.currentCluster, node));
+        });
+        var p = $q.all(promises);
+        p.then(function (responses) {
+          for (var i = 0; i < responses.length; i++) {
```

Comment:
Instead of for loop, use `angular.forEach` -> it is less prone to mistakes, and it is also more clear what is happening: you are iterating through all elements.